lib/web_ui/dev/test_platform.dart

```diff
@@ -164,33 +182,61 @@ class BrowserPlatform extends PlatformPlugin {
-        } else {
-          final File failedFile = File(p.join(file.parent.path, '.${p.basename(file.path)}'));
-          failedFile.writeAsBytesSync(bytes, flush: true);
-
-          // TODO(yjbanov): do not fail Cirrus builds. They currently fail because Chrome produces
-          //                different pictures. We need to pin Chrome versions and use a fuzzy image
-          //                comparator.
-          if (Platform.environment['CIRRUS_CI'] == 'true') {
-            return 'OK';
+
+    // Migrate to flutter_test/lib/src/_goldens_io.dart ?
+    Image golden = decodePng(goldenBytes);
+    Image output = decodePng(bytes);
+    int goldenWidth = golden.width;
+    int goldenHeight = golden.height;
+    int wrongPixels = 0;
+    int numPixels = goldenWidth * goldenHeight;
+    Image diff = Image(goldenWidth, goldenHeight);
+
+    if (goldenWidth == output.width && goldenHeight == output.height) {
+      for(int y = 0; y < goldenHeight; y++) {
+        for (int x = 0; x < goldenWidth; x++) {
+          if (golden.getPixel(x, y) == output.getPixel(x, y)) {
+            diff.setPixel(x, y, Color.fromRgb(255, 255, 255));
+          } else {
+            diff.setPixel(x, y, Color.fromRgb(255, 0, 255));
+            wrongPixels++;
           }
+        }
+      }
+    } else {
+      // Images are completely different resolutions. Bail out big time.
+      wrongPixels = numPixels;
+    }
 
-          return '''
-  Golden file ${file.path} did not match the image generated by the test.
+    if (wrongPixels > 0) {
+      if (write) {
+        file.writeAsBytesSync(encodePng(output), flush: true);
+        return 'Golden file $filename was updated. You can remove "write: true" in the call to matchGoldenFile.';
+      } else {
+        final File failedFile = File(p.join(file.parent.path, '${p.basenameWithoutExtension(file.path)}.out${p.extension(file.path)}'));
+        failedFile.writeAsBytesSync(encodePng(output), flush: true);
+
+        final File failedDiff = File(p.join(file.parent.path, '${p.basenameWithoutExtension(file.path)}.diff${p.extension(file.path)}'));
+        failedDiff.writeAsBytesSync(encodePng(diff), flush:true);
+
+        // TODO(yjbanov): do not fail Cirrus builds. They currently fail because Chrome produces
+        //                different pictures. We need to pin Chrome versions and use a fuzzy image
+        //                comparator.
+        if (Platform.environment['CIRRUS_CI'] == 'true') {
+          return 'OK';
+        }
 
-  The generated image was written to ${failedFile.path}.
+        return '''
+Golden file ${file.path} did not match the image generated by the test (${((wrongPixels/numPixels) * 100).toStringAsFixed(2)}% different pixels).
 
-  To update the golden file call matchGoldenFile('$filename', write: true).
-  ''';
-        }
+* Output image: ${failedFile.path}
+*  Diff pixels: ${failedDiff.path}
+
+To update the golden file call matchGoldenFile('$filename', write: true).
```

Comment:
Shouldn't this be a command-line parameter for the test runner? I feel:  `felt --updateGoldens` is better than having to tweak the code of matchGoldenFile.

(Also, if a golden doesn't exist the first time, we should just create it instead of bailing out of the test. The user will see the new golden file added in their version control status)